kubectl-rook-ceph.sh

```diff
@@ -16,33 +16,269 @@
+        val_exists "$VAL" || fail_error "Flag '$FLAG' does not specify a value"
+        ;;
+      --*) # long flag without a value, e.g., '--help' or '--namespace my-ns'
+        FLAG="$arg"
+        VAL=""
+        ;;
+      -*) # short flags
+        if [[ "${#arg}" -eq 2 ]]; then # short flag without a value, e.g., '-h' or '-n my-ns'
+          FLAG="$arg"
+          VAL=""
+        else # short flag with a value, e.g., '-nmy-ns', or '-n=my-ns'
+          FLAG="${arg:0:2}" # first 2 chars
+          VAL="${arg:2:${#arg}}" # remaining chars
+          VAL="${VAL#*=}" # strip first equal from the value
+        fi
+        ;;
+      *)
+        # This is not a flag, so stop parsing and return the stored remaining args
+        REMAINING_ARGS=("$arg" "$@") # store remaining args BEFORE shifting so we still have the
+        break
+        ;;
+    esac
+    is_flag "$VAL" && fail_error "Flag '$FLAG' value '$VAL' looks like another flag"
+    # run the command with the current value, which may be empty
+    if ! $set_value_function "$FLAG" "$VAL"; then
+      # the flag needs a value, so grab the next arg to use as the value
+      VAL="$1" || fail_error "Could not get value for flag '$FLAG'"
+      shift
+      # fail if the next arg looks like a flag and not a value
+      is_flag "$VAL" && fail_error "Flag '$FLAG' value '$VAL' looks like another flag"
+      # fail because the flag needs a value and value given is empty, e.g., --namespace ''
+      val_exists "$VAL" || fail_error "Flag '$FLAG' does not specify a value"
+      # run the command again with the next arg as its value
+      if ! $set_value_function "$FLAG" "$VAL"; then
+        fail_error "Flag '$FLAG' must have a value" # probably won't reach this, but just in case
+      fi
     fi
+  done
+}
+
+# call this at the end of a command tree when there should be no more inputs past a given point.
+# Usage: end_of_command_parsing "$@" # where "$@" contains the remaining args
+function end_of_command_parsing () {
+  if [[ "$#" -gt 0 ]]; then
+    fail_error "Extraneous arguments at end of input: $*"
+  fi
+}
+
+####################################################################################################
+# 'kubectl rook-ceph help' command
+####################################################################################################
+
+function run_help_command () {
+  end_of_command_parsing "$@" # end of command tree
+  print_usage
+}
+
+####################################################################################################
+# 'kubectl rook-ceph ceph ...' command
+####################################################################################################
+
+function run_ceph_command () {
+  # do not call end_of_command_parsing here because all remaining input is passed directly to 'ceph'
+  kubectl --namespace "$NAMESPACE" exec deploy/rook-ceph-operator -- /bin/bash -c " ceph ${*} --conf=/var/lib/rook/rook-ceph/rook-ceph.config"
```

Comment:
Remind me why this is `/bin/bash -c "ceph $* --conf=...` instead of `ceph --conf=... "$@"` (why /bin/bash, and why --conf arg after the rest of the command?

Reply:
`--conf` is to point to the ceph config file which is required to run ceph commands and I think when we run any command inside pod we pass `bash/sh -c`(`-c` for the command).

Reply:
Let me rephrase:
1. Why is `--conf` the last argument and not the first?
2. Why is `/bin/bash -c` necessary at all? This works fine for me and is less complex: `kubectl exec deploy/rook-ceph-operator -- ceph --conf=... <command>`

Reply:
> Let me rephrase:
> 
> 1. Why is `--conf` the last argument and not the first?
I liked in this way `ceph <some command> --conf=....`
> 2. Why is `/bin/bash -c` necessary at all? This works fine for me and is less complex: `kubectl exec deploy/rook-ceph-operator -- ceph --conf=... <command>`
right, I have removed `/bin/bash -c`